sentry/src/main/java/io/sentry/Baggage.java

```diff
@@ -0,0 +1,173 @@
+
+    if (headerValues != null) {
+      for (final @Nullable String headerValue : headerValues) {
+        final Map<String, String> keyValuesToAdd =
+            extractKeyValuesFromBaggageString(headerValue, logger);
+        keyValues.putAll(keyValuesToAdd);
+      }
+    }
+
+    return new Baggage(keyValues, logger);
+  }
+
+  public static Baggage fromHeader(@Nullable String headerValue, @NotNull ILogger logger) {
+    final Map<String, String> keyValues = extractKeyValuesFromBaggageString(headerValue, logger);
+    return new Baggage(keyValues, logger);
+  }
+
+  private static Map<String, String> extractKeyValuesFromBaggageString(
+      @Nullable String headerValue, @NotNull ILogger logger) {
+    final @NotNull Map<String, String> keyValues = new HashMap<>();
+
+    if (headerValue != null) {
+      try {
+        // see https://errorprone.info/bugpattern/StringSplitter for why limit is passed
+        final String[] keyValueStrings = headerValue.split(",", -1);
+        for (final String keyValueString : keyValueStrings) {
+          try {
+            final String[] keyAndValue = keyValueString.split("=", -1);
+            if (keyAndValue.length == 2) {
+              final String key = keyAndValue[0].trim();
+              final String keyDecoded = decode(key);
+              final String value = keyAndValue[1].trim();
+              final String valueDecoded = decode(value);
+
+              keyValues.put(keyDecoded, valueDecoded);
+            }
+          } catch (Exception e) {
+            logger.log(
+                SentryLevel.ERROR, e, "Unable to decode baggage key value pair %s", keyValueString);
+          }
+        }
+      } catch (Exception e) {
+        logger.log(SentryLevel.ERROR, e, "Unable to decode baggage header %s", headerValue);
+      }
+    }
+    return keyValues;
+  }
+
+  public Baggage(@NotNull ILogger logger) {
+    this(new HashMap<>(), logger);
+  }
+
+  public Baggage(@NotNull Map<String, String> keyValues, @NotNull ILogger logger) {
+    this.keyValues = keyValues;
+    this.logger = logger;
+  }
+
+  public @NotNull String toHeaderString() {
+    final StringBuilder sb = new StringBuilder();
+    String separator = "";
+    int listMemberCount = 0;
+
+    final Set<String> keys = new TreeSet<>(keyValues.keySet());
+    for (final String key : keys) {
+      final @Nullable String value = keyValues.get(key);
+
+      if (value != null && listMemberCount < MAX_BAGGAGE_LIST_MEMBER_COUNT) {
+        try {
+          final String encodedKey = encode(key);
+          final String encodedValue = encode(value);
+          final String encodedKeyValue = separator + encodedKey + "=" + encodedValue;
+
+          final int valueLength = encodedKeyValue.length();
+          final int totalLengthIfValueAdded = sb.length() + valueLength;
+          if (totalLengthIfValueAdded > MAX_BAGGAGE_STRING_LENGTH) {
+            logger.log(
+                SentryLevel.ERROR,
```

Comment:
What log level should this be?
Also should we add the value to the log message?

Reply:
It might break tracing across systems.

Reply:
Might also break dynamic sampling.